### PR TITLE
Fixes adjustment of max_width within macros

### LINKED ISSUE #5404: Cannot set granular width configuration settings to 100% if `max_width = 100`
Thanks for your work on rustfmt!

The granular width configuration settings are documented to be percentages of `max_width` (e.g. [chain_width](https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#chain_width)). However when `max_width` is set to 100 setting their values to `100` causes a warning
```
`fn_call_width` cannot have a value that exceeds `max_width`. `fn_call_width` will be set to the same value as `max_width`
```

The error message is both surprising and incorrect since `fn_call_width = 100` is _not_ a value that exceeds `max_width`.

This does not happen if `max_width` is 99 or 101.

The exact config options that trigger this bug are:
```
max_width = 100
use_small_heuristics = "Off"
fn_call_width = 100
```

```
$ rustfmt --version
rustfmt 1.4.38-stable (fe5b13d 2022-05-18)
```
<hr>
Edit by @ytmimi (add a minimal reproducible code snippet):

```rust
macro_rules! foo {
    () => {};
}
```

### PR DESCRIPTION
**Pull request by csmulhern**
Fixes #5404.

Minimal example:
```rust
macro_rules! foo {
    () => {
        println!("Hello, world!");
    };
}
```

Without this change:
```sh
> cargo run --bin rustfmt -- --config-path=/dev/null --config=max_width=100 --config=fn_call_width=100 ~/Desktop/example.rs
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/rustfmt --config-path=/dev/null --config=max_width=100 --config=fn_call_width=100 /Users/cameron/Desktop/example.rs`
`fn_call_width` cannot have a value that exceeds `max_width`. `fn_call_width` will be set to the same value as `max_width`
```

With this change:
```sh
> cargo run --bin rustfmt -- --config-path=/dev/null --config=max_width=100 --config=fn_call_width=100 ~/Desktop/example.rs
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.03s
     Running `target/debug/rustfmt --config-path=/dev/null --config=max_width=100 --config=fn_call_width=100 /Users/cameron/Desktop/example.rs`
```

### CODE REVIEW

**Comment by csmulhern** [comment]
@ytmimi, could you take a look? Happy to make changes if necessary.

**Comment by csmulhern** [comment]
@Manishearth I see you merging some recent PRs. Do you know who can take a look at this?

**Comment by Manishearth** [comment]
@csmulhern So we're trying to help spread out the maintainer burden, but there are a _lot_ of open PRs right now and we're still triaging them. Open PRs which change formatted code are tricky because it needs to be determined if it's a rbeaking change or not.

We'll get to them eventually, but no promises.

It will help if you show a before and after of the formatted code.

**Comment by Manishearth** [comment]
It _seems_ like from the PR description that this doesn't actually impact formatting, this makes certain previously-unintentionally-invalid values of a config start working again.

**Comment by csmulhern** [comment]
Yes, this (mostly*) shouldn't impact formatting.

Today, we have a bunch of properties that control breakpoints for how long certain lines can be. This is primarily `max_width`, but also [fn_call_width](https://rust-lang.github.io/rustfmt/#fn_call_width), [attr_fn_like_width](https://rust-lang.github.io/rustfmt/#attr_fn_like_width), [struct_lit_width](https://rust-lang.github.io/rustfmt/#struct_lit_width), [struct_variant_width](https://rust-lang.github.io/rustfmt/#struct_variant_width), [array_width](https://rust-lang.github.io/rustfmt/#array_width), [chain_width](https://rust-lang.github.io/rustfmt/#chain_width), [single_line_if_else_max_width](https://rust-lang.github.io/rustfmt/#single_line_if_else_max_width). By default, these are all derived from `max_width`, due to `use_small_heuristics` defaulting to true.

In order to support formatting in macros, it looks like rustfmt just trims down `max_width` by the indentation of the macro arm body, and then runs formatting of the body with that adjusted config. The problem is, those other `max_width` style properties are not adjusted. So then when those values are read inside the formatting for the macro body, they can be larger than `max_width`, which triggers [this error](https://github.com/rust-lang/rustfmt/blob/6ac1fcd7a3336a57793206d3a447f0772e76e2cf/src/config/config_type.rs#L437), and the values are clamped down to `max_width`. Note that if these values are unset, there's no issue, because it will just default to the heuristic value.

This change makes the reduction based on the indent symmetric across all of the max_width style properties, and adds a new method to ensure this is done universally.

\* The edge case where this would affect formatting would be if e.g. `max_width` was set to 80, and `fn_call_width` was set to 76, and the macro body indent was 8. In today's world, `fn_call_width` in the macro body would be larger than the `max_width` within the macro body (72), and would get clamped down to 72. With this fix, both would be reduced by 8 (`max_width` 80 -> 72, `fn_call_width` 76 -> 68).

**Comment by csmulhern** [comment]
@Manishearth / @ytmimi, any feedback on this?

**Comment by csmulhern** [comment]
@jieyouxu do you know what the status is re: if these sort of PRs are ready for review? Is there anything I can do to help make this more reviewable?

**Comment by jieyouxu** [comment]
> Is there anything I can do to help make this more reviewable?

It's largely (rustfmt) review bandwidth that's fairly scarce. I can do a re-review this weekend.

File: src/macros.rs
Comment (jieyouxu) [inline src/macros.rs:1]: Question:

> * The edge case where this would affect formatting would be if e.g. `max_width` was set to 80, and `fn_call_width` was set to 76, and the macro body indent was 8. In today's world, `fn_call_width` in the macro body would be larger than the `max_width` within the macro body (72), and would get clamped down to 72. With this fix, both would be reduced by 8 (`max_width` 80 -> 72, `fn_call_width` 76 -> 68).

Is there a (albeit synthetic) test that can demonstrate the *after* effect (of this edge case)?

File: src/config/config_type.rs
Comment (jieyouxu) [inline src/config/config_type.rs:1]: Remark: as I understand it (like mentioned previously), the formatting impact *should* be impacts stable options, but previously warning (and clamped to `max_width`) -> now actually formatted

File: src/macros.rs
Comment (jieyouxu) [inline src/macros.rs:1]: Remark (not for this PR): hm, these `eprintln!`s feel problematic to me in terms of being able to actually *test* them. We probably need to not directly use `eprintln!` and instead use some other custom interceptor which allows promoting warnings -> errors when testing 🤔

Comment (jieyouxu) [inline src/macros.rs:1]: Suggestion: could you add a snapshot test for one or more of the MCVE examples in https://github.com/rust-lang/rustfmt/issues/5404 (potentially as `target` tests only) to demonstrate that even with the set of changes here, under the settings that the examples are *not* formatted differently?

(I hope I'm understanding this correctly; please do correct me if I am misunderstanding.)

**Review by jieyouxu — COMMENTED** [review]
Hi, thanks for the PR and sorry for the slow response. I have a few questions/suggestions.

@rustbot author

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustfmt/6651/7991944d54b962ac09e5ba67b76635cf45ce069e..da6befc1ee3e99fe207ca795aba8df08324ec6bd)*

**Comment by jieyouxu** [comment]
(Also sorry about the force-push, initially I thought that I needed to run DiffCheck, but then it occurred to me the effect of the changes here is basically not going to show up in the corpus AIUI.)

File: src/macros.rs
Comment (csmulhern) [reply src/macros.rs:1]: I've added a prior commit in the chain that adds test fixtures for these macro use cases, with both default heuristics, and manually crafted widths that will trigger this edge case when the macro body indent exceeds the difference between the custom width and `max_width`.

Note that the error log currently happens _whenever_ the "clipping" behavior occurred, which is for any file with a macro body if you have custom width values close to / equal to `max_width`.

This change fixes the error logging, but it only effects formatting when the length of a line is specifically within the gap between `max_width - <custom width>` and `max_width - <macro body indent>`.

The `long` variant of these newly added tests is the macro arm specifically crafted to trigger these conditions. You'll notice that the `long` variant is the same in `source` and `target` before the commit.

When it comes to this being formatting affecting:
1) I believe these edge cases _should_ be formatted the way they are post-commit. The user expressed for e.g. `fn_call_width` to be less than `max_width`, and so clipping to `max_width` feels inappropriate.
2) `format_macro_bodies` is an unstable option, so I believe formatting changes should be tolerated here regardless of compatibility.

File: src/config/config_type.rs
Comment (csmulhern) [reply src/config/config_type.rs:1]: My comment in the thread above has hopefully addressed this.

File: src/macros.rs
Comment (csmulhern) [reply src/macros.rs:1]: Agreed!

Comment (csmulhern) [reply src/macros.rs:1]: Done (more details in the thread above).

Comment (jieyouxu) [reply src/macros.rs:1]: Thanks for the detailed explanation, this is what I was having trouble figuring out 👍 

> [!NOTE]
>
> I consider this review chain **resolved**, but intentionally leaving open for other reviewers for context.

Comment (jieyouxu) [reply src/macros.rs:1]: Tracked elsewhere in https://github.com/rust-lang/rustfmt/issues/6933

File: src/config/config_type.rs [hunk 1]
```diff
@@ -423,6 +423,63 @@ macro_rules! create_config {
...
+            /// configuration for a particular formatting context in which the
+            /// total available width needs to be reduced. This reduces the
+            /// size of max_width, and all other properties affected by small
+            /// heuristics, without treating those adjustments as overrides.
+            ///
+            /// As an example use case, this method is used when formatting
+            /// arms within a declarative macro.
+            #[allow(unreachable_pub)]
+            pub fn reduce_max_width(&mut self, delta: usize) {
+                self.adjust_max_width(-1 * delta as isize);
```
Comment (jieyouxu) [inline src/config/config_type.rs:438]: Remark: in theory, if we want to be very careful around handling of large widths, this has a potential problem where `delta`'s `usize -> isize` cast can wraparound for sufficiently large `usize`, namely `[isize::MAX + 1, usize::MAX]`.

However, I *don't* think this is a blocking concern for *this* PR and should not be something that's run into super frequently in practice (esp. if host pointer width is 64+). AFAICT in other parts of the code base we are also not *super* careful about large widths, so even if this wraps around there are other parts of the codebase that will have similar problems, AFAIK.

**Review by jieyouxu — APPROVED** [review]
Thanks, the changes look good to me. I have one non-blocking observation re. large width behavior, but that is not something that I think should be addressed in *this* PR. I'll hand this review over for a final review by one of the maintainers.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustfmt/6651/7991944d54b962ac09e5ba67b76635cf45ce069e..947892b1653ebfcfdf7a60559c908c27a9c14397)*

**Comment by csmulhern** [comment]
Thanks for the review, @jieyouxu. Do you know how we can get a maintainer to take a look?

**Comment by jieyouxu** [comment]
> Do you know how we can get a maintainer to take a look?

Maintainers are unfortunately very busy (and bandwidth is very limited versus amount of PRs). They will get to this PR, just that it might take some time.

File: src/macros.rs
Comment (jieyouxu) [inline src/macros.rs:1]: Remark: I don't expect this to impact anything (in theory this affects the unstable option only), but just in case: [DiffCheck @ E2024 x SE2024](https://github.com/rust-lang/rustfmt/actions/runs/30898412508) 🍏.

(Not that this is very representative.)

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustfmt/6651/c648a61616c6be0ab0d391709ebab77114d42e24..012d3f587fe274f987bd9ac906f3c2acc3cf6144)*

**Comment by jieyouxu** [comment]
I force-pushed fixes for two tiny nits I had remaining. There's of course still the usual integer wrapping stuff, but that's not just this PR, we just don't handle those *super* carefully in general.

**Review by jieyouxu — APPROVED** [review]
Sorry for the delay, I think this looks good to merge.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustfmt/6651/7bc6cd796f03a0bc5299c66fd6a91e6c32c3fd78..93e78f2d3a0901718530aa871565257a41869983)*
